src/Xrefcheck/Verify.hs

```diff
@@ -241,20 +246,43 @@ instance Given ColorMode => Buildable VerifyError where
       #{err}
       |]
 
-    PermanentRedirectError url Nothing ->
+    RedirectChainCycle chain ->
       [int||
-      Permanent redirect found:
-      #{url}
+      Cycle found in the following redirect chain:
+      #{interpolateIndentF 2 $ attachToRedirectChain chain "here"}
       |]
 
-    PermanentRedirectError url (Just redirectedUrl) ->
+    RedirectChainBroken chain ->
       [int||
-      Permanent redirect found. Perhaps you want to replace the link:
-        #{url}
-      by:
-        #{redirectedUrl}
+      Missing location header in the following redirect chain:
+      #{interpolateIndentF 2 $ attachToRedirectChain chain "no location header"}
       |]
 
+    RedirectChainLimit chain ->
+      [int||
+      The follow redirects limit has been reached in the following redirect chain:
+      #{interpolateIndentF 2 $ attachToRedirectChain chain "stopped before this one"}
+      |]
+
+    RedirectRuleError chain mOn ->
+      [int||
+      #{redirect} found:
+      #{interpolateIndentF 2 $ attachToRedirectChain chain "stopped before this one"}
+      |]
+      where
+        redirect :: Text
+        redirect = case mOn of
+          Nothing -> "Redirect"
+          Just RROPermanent -> "Permanent redirect"
+          Just RROTemporary -> "Temporary redirect"
+          Just (RROCode code) -> show code <> "redirect"
```

Comment:
Don't we need a space between the code and "redirect"? `show code <> " redirect"`

Reply:
True! I have no Golden tests with these messages because they would require to also have a mock server to replicate very specific redirect cases.